src/schema.rs

```diff
@@ -46,13 +47,15 @@ pub enum Schema {
     /// A `string` Avro schema.
     /// `String` represents a unicode character sequence.
     String,
-    /// A `array` Avro schema. Avro arrays are required to have the same type for each element.
-    /// This variant holds the `Schema` for the array element type.
-    Array(Box<Schema>),
+    /// A reference to a type defined in this schema
+    TypeReference(NameRef),
+    /// A `array` Avro schema.
+    /// `Array` holds a counted reference (`Rc`) to the `Schema` of its items.
+    Array(Arc<Schema>),
     /// A `map` Avro schema.
     /// `Map` holds a pointer to the `Schema` of its values, which must all be the same schema.
     /// `Map` keys are assumed to be `string`.
-    Map(Box<Schema>),
+    Map(Arc<Schema>),
```

Comment:
I am also curious as to why `Array` and `Map` now need an `Arc`, instead of a simple `Box`. Would you mind elaborating a bit? 🙂